syntax = "proto3";

package planning;

import "common.proto";
import "google/protobuf/timestamp.proto";
import "google/protobuf/field_mask.proto";

enum PaymentStatus {
    PAYMENT_STATUS_UNKNOWN=0;
    // Payment plan is in good standing and all payments are current
    PAYMENT_STATUS_CURRENT=1;
    // This plan is fully paid
    PAYMENT_STATUS_COMPLETED=2;
    // The user has requested this payment to be cancelled
    PAYMENT_STATUS_CANCELLED=3;
    // The payment plan is not current because the user has missed a payment and still have not paid it
    PAYMENT_STATUS_IN_DEFAULT=4;
}

enum PaymentActionStatus {
    PAYMENT_ACTION_STATUS_UNKNOWN = 0;
    // Payment is pending
    PAYMENT_ACTION_STATUS_PENDING = 1;
    // payment is completed
    PAYMENT_ACTION_STATUS_COMPLETED = 2;
    // payment defaulted
    PAYMENT_ACTION_STATUS_IN_DEFAULT = 3;
}

message PaymentAction {
    // account to which payment shall be made
    string account_id = 1;
    // amount of payment
    float amount = 2;
    // Planned date of transaction
    google.protobuf.Timestamp transaction_date = 3;
    // status of payment action
    PaymentActionStatus status = 4;
}

message PaymentPlan {
    string payment_plan_id = 1;
    string user_id = 2;
    repeated string payment_task_id = 3;
    float timeline = 4;
    PaymentFrequency payment_freq = 5;
    float amount_per_payment = 6;
    PlanType plan_type = 7;
	// Expected date the plan should be completed
	google.protobuf.Timestamp end_date=8;
	// This Payment plan has not completed and is still active
	bool active=9;
	// Current status of this plan
	PaymentStatus status=10;
    // payment actions of the plan
    repeated PaymentAction payment_action = 11;
}

// CRUDE Methods
message GetPaymentPlanRequest {
    string payment_plan_id = 1;
}

message ListPaymentPlanRequest {}

message ListPaymentPlanResponse {
    repeated PaymentPlan payment_plans = 1;
}

message UpdatePaymentPlanRequest {
    string payment_plan_id = 1;
    PaymentPlan payment_plan = 2;
    google.protobuf.FieldMask mask = 3;
}

message DeletePaymentPlanRequest {
    string payment_plan_id = 1;
}

message DeletePaymentPlanResponse {
    DELETE_STATUS status = 1;
    PaymentPlan payment_plan = 2;
}